hcl2template/types.datasource.go

```diff
@@ -101,19 +102,48 @@ func (cfg *PackerConfig) startDatasource(dataSourceStore packer.DatasourceStore,
 			Severity: hcl.DiagError,
 		})
 	}
+
+	// HACK:
+	// This is where we parse the variables being used in the data sources.
+	// By passing in the DatasourceContext variable, we tell the EvalContext
+	// that since this is a datasource being evaluated, we should not allow
+	// other data sources to be decoded into it. When secondaryEvaluation is
+	// true, we know that this data source needs another data source in order
+	// to be evaluated. So we instead retrieve a different EvalContext.
+	// This is a brute force method to enable data sources to depend on each
+	// other, and a more elegant solution will be available once we implement a
+	// true DAG for Packer.
+	var decoded cty.Value
+	var moreDiags hcl.Diagnostics
 	body := block.Body
-	decoded, moreDiags := decodeHCL2Spec(body, cfg.EvalContext(DatasourceContext, nil), datasource)
+	if secondaryEvaluation {
+		// LocalContext is a lie! See above.
+		decoded, moreDiags = decodeHCL2Spec(body, cfg.EvalContext(LocalContext, nil), datasource)
+	} else {
+		decoded, moreDiags = decodeHCL2Spec(body, cfg.EvalContext(DatasourceContext, nil), datasource)
+	}
+
 	diags = append(diags, moreDiags...)
 	if moreDiags.HasErrors() {
-		return nil, diags
+		for _, err = range moreDiags.Errs() {
+			// If the error is just that there's no "data" object in the
+			// context, don't fail. We will track this data source for decoding
+			// again later, once we've evaluated all of the datasources.
+			// return nil, diags
+			if !strings.Contains(err.Error(), `There is no variable named "data"`) {
+				// There's an error that isn't just a recursive data source
+				// interpolation error
+				return nil, diags
+			}
```

Comment:
I'd probably refrain from checking for error messages like this, the messages in the HCL/CTY libraries often change.

Reply:
Ah, I don't need this check anyway since I know ahead of time whether this actually needs to have the data variable in context. Removing it.

Reply:
nice !